Various Samfire fixes

to266 opened the pull request:
- solves both issues in https://github.com/hyperspy/hyperspy/issues/1347
 - a more robust plotting of histograms
 - enables updating parameter boundaries between samfire runs
 - enables updating GOF test between runs
 - removes things form public interface that should not have been there in the first place

### API change
`samfire_pool.update_optional_names` got renamed to `samfire_pool.update_parameters`. Also no longer updates just the optional names, but other worker parameters as well (e.g. parameter boundaries, goodness test)

to266 committed:
fix goodness_test updating in workers

to266 committed:
samfire fixes

to266 committed:
rename gt_dump -> _gt_dump

to266 committed:
do not add too many jobs, always pass chisq

to266 committed:
testing samfire more general passing methods

to266 committed:
not overwrite normal data

to266 committed:
parameter boundaries, logging more careful data passing

to266 committed:
prevent workers copying unnecessary things

to266 committed:
not doing anything automatically when slicing models

to266 committed:
Merge remote-tracking branch 'upstream/RELEASE_next_patch' into ENH_samfire

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -1722,12 +1722,12 @@ def __getitem__(self, slices):
-                auto_add_edges=False)
-        else:
-            _model = _signal.create_model()
+        # let's not do anything automatically, since then it's difficult to keep
```
I think that this is a dangerous approach that works with the code as it is at present but may break things in the future. I think that auto adding components is a bad idea anyway so, what about checking if it is an eds/eels model (the only two ones that create components so far) and if they are, set the auto keywords explicitly to `False`. In parallel we could open an issue to remove the `auto*` keys in favour of methods that the use have to run explicitly.

to266 replied:
I would rather move this to `next_minor` and at the same time remove autoadding anything. As you can see it used to be checking if it's an EELS model, etc., but then EDS model was added and hence the bug. In general I don't like adding exceptions where either a general approach or sane defaults should suffice...

In this case the "sane defaults" might include "sane kwarg names", making the current approach passable ;)

francisco-dlp replied:
I agree with the principle. However, to change the API we need to wait for `next_major` which may be a too long wait to get all of this fixes...

to266 committed:
no X for workers needed. Try doing this explicitly

to266 commented:
Also it seems I __may__ have finally fixed the appveyor samfire tests. Almost not surprisingly it seems to have been a matplotlib / traits / traitui error

francisco-dlp commented on hyperspy/samfire.py:
```diff
@@ -483,15 +491,26 @@ def _next_pixels(self, number):
-        for comp_name, comp in dic['components'].items():
+            d_ind = tuple([0 for _ in dict_['dof.data'].shape])
+        m = self.model
+        for k in dict_.keys():
```
The commented out code below is more readable than the more compact and general new one. Is the extra generality needed? If yes, what about leaving a comment explaining what the code does instead or in addition to leaving the old code commented out below?

to266 replied:
I'll add an explanation.

In particular, I tried adding `corr` to a `Model2D` - a correlation measure to be used instead of `red_chisq`, and it was troublesome adding this to samfire.

In this case I coded the thing in the class (not included in this PR though), but generally it could be anything that gets computed on fitting a pixel, since there now is a `fitted` event that may trigger any arbitrary calculations that save to the model itself.

francisco-dlp commented on hyperspy/samfire_utils/samfire_pool.py:
```diff
@@ -51,8 +51,8 @@ class SamfirePool(ParallelPool):
         queues.
-    update_optional_names
-        updates the optional names in all workers
+    update_parameters
```
This seems like an API change...

to266 replied:
Huh, true.... 
On the other hand, I doubt anyone else was aware of the function. On the yet another hand, this "bug" is actually __very__ bad if you do anything else that what I was used to - not only you can't change the parameter boundaries, but also the GOF test, other things... Your call anyway. I can PR to the major branch just as well

francisco-dlp replied:
I am shamefully willing to turn a blind eye to this API change (hopefully @vidartf is not watching...)

francisco-dlp replied:
However, it'll be good to describe in the header and possibly the changes list just in case somebody actually cares...

to266 committed:
Generalising worker->master signal data passing.

Automatically copies signal data (i.e. "*.data") fields form worker-passed
dictionary to appropriate signals. This allows for user adding arbitrary
computations one `fitted` event is triggered and storing the results in a
signal similar to `dof` or `red_chisq`, etc.

to266 committed:
API:update_optional_names -> update_parameters doc

Updates not just optional names, but other samfire parameters as well,
currently:
 - gootness test
 - optional component names
 - parameter boundaries

to266 committed:
TODO msg for next major release

to266 commented:
Added some comments, updated PR header, created an issue to remind us to make the appropriate API change for the next major version... Anything else?
